src/net.c

```diff
@@ -1363,13 +1402,28 @@ coap_send_internal(coap_session_t *session, coap_pdu_t *pdu) {
   }
 #endif /* !COAP_DISABLE_TCP */
 
-  bytes_written = coap_send_pdu( session, pdu, NULL );
+#ifdef HAVE_OSCORE
+  if (session->oscore_encryption &&
+      !(pdu->type == COAP_MESSAGE_ACK && pdu->code == COAP_EMPTY_CODE)) {
+    /* Refactor PDU as appropriate RFC8613 */
+    coap_pdu_t *osc_pdu = coap_oscore_new_pdu_encrypted(session, pdu, NULL,
```

Comment:
It looks like the sequence number is not incremented when retransmitting. The OSCORE RFC does not seem to mandate incrementing the sequence number in retransmissions, but not doing so ensues problems. When an ACK to a CON gets lost, the sender's retransmissions will be discarded as they appear replayed. Besides, if a CON gets lost and a lot of other messages arrive in the meantime, the replay window may be so short that once the CON arrives it gets discarded:

A -x B: CON with sequence number n
A -> B: NON with sequence number n + 1
A -> B: NON with sequence number n + 2
...
A -> B: CON with sequence number n
B: Discard CON since n is no longer in the replay window

I therefore do the OSCORE securing in `coap_send_pdu`.

Reply:
Interesting point - thanks for doing the review.  I had not thought about that.  Doing the OSCORE encryption work in `coap_send_pdu() `would mean that whenever `coap_retransmit()` or similar gets called, the sender sequence number will get increased.

I will look at getting in a fix for this.

Reply:
Code updated and pushed.  The OSCORE encryption had to be done in `coap_session_send_pdu()` to properly handle any re-transmits.

Reply:
Incrementing the pIV (seq no) when re-transmitting a PDU can be problematic.  Lets say the ACK is delayed, the the re-transmitted PDU is sent and then the first ACK arrives.  At this point, the `external_aad `uses which sending pIV - the first transmission, or the re-transmission?

It has to use the pIV of the initial transmission - otherwise there is a decryption failure.